### ISSUE TITLE
Fatal error when no product is passed to the pre-orders

### ISSUE THREAD

**Issue by mikkamp** [skyverge/wc-plugin-framework#352]
We are seeing the following fatal error:

```
2019-10-22T11:24:42+00:00 CRITICAL Uncaught Error: Call to a member function is_type() on bool in /wp-content/plugins/woocommerce-pre-orders/includes/class-wc-pre-orders-product.php:257
Stack trace:
#0 /wp-content/plugins/woocommerce-gateway-authorize-net-cim/vendor/skyverge/wc-plugin-framework/woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-pre-orders.php(98): WC_Pre_Orders_Product::product_is_charged_upon_release(false)
#1 [internal function]: SkyVerge\WooCommerce\PluginFramework\v5_4_3\SV_WC_Payment_Gateway_Integration_Pre_Orders->maybe_force_tokenization(false)
#2 /wp-includes/class-wp-hook.php(288): call_user_func_array(Array, Array)
#3 /wp-includes/plugin.php(208): WP_Hook->apply_filters(false, Array)
#4 /wp-content/plugins/woocommerce-gateway-authorize-net-cim/vendor/skyverge/wc-plugin-framework/woocommerce/payment-gateway/payment-tokens/class-sv-wc-payment-gateway-p in /wp-content/plugins/woocommerce-pre-orders/includes/class-wc-pre-orders-product.php on line 257
```

This happens when the function `product_is_charged_upon_release` without a product. See:
https://github.com/skyverge/wc-plugin-framework/blob/5.4.3/woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-pre-orders.php#L98

It initially checks the order with (which returns true):
`WC_Pre_Orders_Order::order_contains_pre_order( $order_id )`
But then fails to find a product with:
`WC_Pre_Orders_Order::get_pre_order_product( $order_id )`

The reason this fails is because the order has already been marked as completed, which according to the docs means it should be treated as: https://docs.woocommerce.com/document/pre-orders/#section-11

> Once a pre-order status is set to “complete” the parent Order is no longer considered to be “pre-ordered” and should be managed as would any other order.

Because of this it doesn't return a product. It's a bit unclear why the first check should still pass, but I've also logged an issue for the Pre-Orders extension to handle this scenario better.

However it would still be nice if there is no product returned to fallback to the default of not forcing tokenization. This will prevent the fatal error from occurring.

**Comment by unfulvio** [comment]
Thanks for the heads up @mikkamp - we'll look into it.

**Comment by unfulvio** [comment]
hey @mikkamp recently released version 5.5.1 of the framework will address this - currently rolling updates in some of our gateways so they will start making use of a newly introduced sanity check as you recommended above - please do get back to us if something is still off -- thanks again!
